apiserver/facades/client/keymanager/keymanager.go

```diff
@@ -61,338 +65,448 @@ func (api *KeyManagerAPI) checkCanWrite(ctx context.Context, sshUser string) err
+	// eventually Juju could support adding keys for individual users. We have
+	// now partially wired this support up in the service and when adding keys
+	// we speicfy the user to which the keys should be added for.
+	//
+	// Currently the Juju client doesn't support adding keys for a specific user
+	// but for the user arg in params the client always sets this to "admin".
+	// We can't rely on this value at the moment and have to use the
+	// authenticated entity.
+
+	if err := api.checkCanRead(ctx); err != nil {
+		return params.StringsResults{}, apiservererrors.ServerError(err)
+	}
 	if len(arg.Entities.Entities) == 0 {
 		return params.StringsResults{}, nil
 	}
 
-	// For now, authorised keys are global, common to all users.
-	cfg, err := api.model.ModelConfig(ctx)
-	if err != nil {
-		// Return error embedded in results for compatibility.
-		// TODO: Change this to a call-error on next facade bump
-		results := transform.Slice(arg.Entities.Entities, func(_ params.Entity) params.StringsResult {
-			return params.StringsResult{Error: apiservererrors.ServerError(err)}
-		})
-		return params.StringsResults{Results: results}, nil
-	}
-	keys := ssh.SplitAuthorisedKeys(cfg.AuthorizedKeys())
-	keyInfo := parseKeys(keys, arg.Mode)
-
-	results := transform.Slice(arg.Entities.Entities, func(entity params.Entity) params.StringsResult {
-		// NOTE: entity.Tag isn't a tag, but a username.
-		if err := api.checkCanRead(ctx, entity.Tag); err != nil {
-			return params.StringsResult{Error: apiservererrors.ServerError(err)}
+	results := make([]params.StringsResult, 0, len(arg.Entities.Entities))
+	for range arg.Entities.Entities {
+		if api.authedUser.Kind() != names.UserTagKind {
+			results = append(results, params.StringsResult{
+				Error: apiservererrors.ParamsError(
+					params.CodeTagKindNotSupported,
+					"authorised user %q unsupported, can only accept tags of kind %q",
+					api.authedUser, names.UserTagKind,
+				),
+			})
+			continue
 		}
-		// All keys are global, no need to look up the user.
-		return params.StringsResult{Result: keyInfo}
-	})
-	return params.StringsResults{Results: results}, nil
-}
 
-func parseKeys(keys []string, mode ssh.ListMode) (keyInfo []string) {
-	for _, key := range keys {
-		fingerprint, comment, err := ssh.KeyFingerprint(key)
-		if err != nil {
-			keyInfo = append(keyInfo, fmt.Sprintf("Invalid key: %v", key))
+		user, err := api.userService.GetUserByName(ctx, api.authedUser.Id())
+		switch {
+		case errors.Is(err, accesserrors.UserNameNotValid):
+			results = append(results, params.StringsResult{
+				Error: apiservererrors.ParamsError(
+					params.CodeUserInvalidName,
+					"invalid user name: %q",
+					api.authedUser.Id(),
+				),
+			})
+			continue
+		case errors.Is(err, accesserrors.UserNotFound):
+			results = append(results, params.StringsResult{
+				Error: apiservererrors.ParamsError(
+					params.CodeUserNotFound,
+					"user %q does not exist",
+					api.authedUser.Id(),
+				),
+			})
```

Comment:
This would allow the leaking of information about a specific user. i.e. you're user A, who can read but can't write, but can ask for random user names to see if user B exists. This is a form of a rainbow attack.

Reply:
It won't because of how badly we messed up this API. You can specify as many users as you like but we only ever lookup the authed user because the client is implemented wrong.

Reply:
You are also authenticated as a model admin at this stage.

Reply:
Can we add a comment to the effect?

Reply:
Done & Done